### PR TITLE
Use `ckan.site_url` to generate urls of resources

### LINKED ISSUE #2592: Download URL has wrong domain name
Based on the mails form the ckan-dev mailinglist (https://lists.okfn.org/pipermail/ckan-dev/2015-August/009236.html), there seems to be a problem when a file is uploaded manually and if the external address of CKAN is different from the internal one.

In our setup the admin users use an internal-only domain to manage the open data portal. The external address is different and set correctly in the configuration (`ckan.site_url`). But when generating the absolute download URL of a resource, this setting is not considered, thus the download gets an internal URL, which an external user can not download.

@amercader already pointed me in the right direction, as where the URL is generated, the code is here: https://github.com/ckan/ckan/blob/master/ckan/lib/dictization/model_dictize.py#L118.

PR follows.

### PR DESCRIPTION
**Pull request by metaodi**
Fixes #2592 on master

### CODE REVIEW

**Comment by wardi** [comment]
@metaodi this is great.

I would prefer if the fix was one level lower though, inside ckan's `url_for` helper. Any time a caller passes qualified=True ckan's helper should pass the protocol and host to the routes url_for function.

**Comment by metaodi** [comment]
@wardi good point, I'll change that :+1:

File: ckan/lib/helpers.py [hunk 1]
```diff
@@ -110,6 +110,17 @@ def url(*args, **kw):
     return _add_i18n_to_url(my_url, locale=locale, **kw)
 
 
+def get_site_url_and_protocol():
```
Comment (wardi) [inline ckan/lib/helpers.py]: This returns (protocol, host). How about renaming this function `get_site_protocol_and_host()`

Also please add a docstring that shows that when for e.g. `site_url = http://example.com/prefix/` this helper returns `('http', 'example.com')`

**Comment by wardi** [comment]
I'll need to test this on a site that has ckan running with a prefix on site_url

**Comment by metaodi** [comment]
@wardi I updated the name and the docstring. About the prefix -> if I'm not wrong, this would not be part of the [site_url](http://docs.ckan.org/en/943-writing-extensions-tutorial/configuration.html#ckan-site-url), but rather the [root_path](http://docs.ckan.org/en/943-writing-extensions-tutorial/configuration.html#ckan-root-path).

**Comment by wardi** [comment]
@metaodi oh, good call. Recent directions on ckan-dev have been saying to use site_url, but root_path does look like the right place for that.

I'll need to test that root_path still works then

**Comment by metaodi** [comment]
@wardi I added 3 tests to make sure `url_for` works as expected.

**Comment by wardi** [comment]
@metaodi would you add one more with a non-default root_path set?

**Comment by metaodi** [comment]
@wardi I added the tests, there seems to something broken, I'll look into that.

**Comment by metaodi** [comment]
@wardi I think I could fix it. The previous implementation of the `root_path` thingy seems broken, there was a misuse of the variables `root` and `root_path`. Please check if it makes sense now. My tests pass, but maybe there are other cases that I don't check.

**Comment by wardi** [comment]
@metaodi very nice! Thanks!

**Pull request merged**